elastic/kibana
# Add ELU history metrics

## Summary

Close https://github.com/elastic/kibana/issues/184836

Adds a new endpoint: `/api/_elu_history` for the purposes of tracking load.

```ts
interface Response {
  /**
   * Event-loop utilization represented as an average of a number of collections as buckets
   * @remark 3 load windows borrows from the `uptime` CLI tool on macOS and Linux, but is not necessarily 1m, 5m, 15m. The
   *         actual time range covered is determined by our collection interval (configured via `ops.interval`, default 5s)
   *         and the number of samples held in each window. So by default short: 15s, medium: 30s and long 60s.
   */
  history: {
    /** The average ELU for the short window */
    short: number;
    /** The average ELU for the medium window */
    medium: number;
    /** The average ELU for the long window */
    long: number;
  };
}
```

## How to test

Start Kibana locally (`yarn start --no-base-path`) and immediately run the script below.

```bash
watch -n1 curl -s http://localhost:5601/api/_elu_history
```

Once Kibana starts responding you should see the `short  > medium > long`, then `short < medium < long` and eventually `short ~= medium ~= long` if you let Kibana idle (basically, long needs to "lag" behind short and medium).

## Questions

1. An alternative implementation exposes this directly via the `/api/stats` endpoint as a new section under `event_loop_utilization`. I'm assuming this approach is preferable.
2. ~Naming... What shall we call this endpoint?~ history > load
